### DISCUSSION TITLE
How to compile and bootstrap from source

### DISCUSSION THREAD

**Discussion by bastien-roucaries** [sbt/sbt#7168] [Q&A] [answered]
Hi,

I want to compile sbt from source for debian

I could not find documentation for compiling from source sbt

Could you please:
- explain how to compile from source
- the bootstrap process

rouca

**Comment by SethTisue** [comment]
To build the current version of sbt, you need to use a previous version of sbt. There is no way around this.

There is some pretty extensive previous discussion at https://contributors.scala-lang.org/t/compiling-scala-and-sbt-for-debian-distro/3620

> explain how to compile from source

there are full instructions in DEVELOPING.md

**Comment by bastien-roucaries** [comment]
So:
- I need a sbt build that is statically compiled => where I should find it ?
- With this statis version I need to build sbt io librarymanagement zinc
- then build a local sbt (step 2)
- rebuild a final sbt (step3)

Now I can do the the build and version by version update for debian.

Other question:
- could it build without internet
- how to locally install module ? And dynamic link module to built sbt

BTW bootstraping is allowed for debian, but it need to pass by contrib distrib

**Comment by SethTisue** [comment, marked as answer]
> I need a sbt build that is statically compiled => where I should find it ?

You don't need to involve any native binaries. sbt runs on the JVM, from JARs not native binaries, and is launched by a shell script. 

(If I recall correctly, the sbt download bundles available from the sbt website include native binaries for sbtn, but you don't actually need sbtn for any of this.)

> could it build without internet

I think so, but you'd need to pre-populate a local Maven repository with the required JARs.

> how to locally install module ? And dynamic link module to built sbt

I don't know what this means.

**Reply by bastien-roucaries** [reply]
> > I need a sbt build that is statically compiled => where I should find it ?
> 
> You don't need to involve any native binaries. sbt runs on the JVM, from JARs not native binaries, and is launched by a shell script.

Yes but does this jar need other jar to run or native binaries. If so I am blocked here. I need in this case a self contained sbt build.
> 
> (If I recall correctly, the sbt download bundles available from the sbt website include native binaries for sbtn, but you don't actually need sbtn for any of this.)
> 
> > could it build without internet
> 
> I think so, but you'd need to pre-populate a local Maven repository with the required JARs.

What is the list of the maven module ? Does this module need sbt ? If so I will be blocked.
> 
> > how to locally install module ? And dynamic link module to built sbt
> 
> I don't know what this means.

Install means copy sbt to /usr/bin/ and library to /usr/share/java/*.jar

**Reply by SethTisue** [reply]
> Install means copy sbt to /usr/bin/ and library to /usr/share/java/*.jar

Okay, and what is your question about that?

**Reply by SethTisue** [reply]
> What is the list of the maven module ?

I'm not sure if anybody has made that list. If not, you could start with an empty Coursier cache, build sbt, and see what ends up in that cache.

**Reply by SethTisue** [reply]
> Does this module need sbt ? If so I will be blocked.

Well, sbt is bootstrapped, some circularity is unavoidable. But as you already wrote,

> bootstrapping is allowed for debian, but it need to pass by contrib distrib

**Comment by eed3si9n** [comment]
Something to be aware is accidental/historical conflation around what "sbt" refers to.
1. sbt can refer to a launcher program named `sbt-launch.jar` (we call this "launcher"). When package managers "install sbt", they aren't actually installing real sbt, they are mostly installing the launcher. This can be downloaded from https://repo1.maven.org/maven2/org/scala-sbt/sbt-launch/1.8.2/sbt-launch-1.8.2.jar, but you can grab it from https://github.com/sbt/sbt/releases/download/v1.8.2/sbt-1.8.2.tgz.
2. sbt can refer to a shell script named `sbt` (we call this "runner"), which calls the launcher, and also included into https://github.com/sbt/sbt/releases/download/v1.8.2/sbt-1.8.2.tgz.
3. sbt can refer to the program (I call this "mother ship" or "main artifact") that does the building. This is published to Maven Central, and needs to be resolved using some resolver, like the launcher.

This mechanism allows users to install the runner + launcher machine-wide, but change the sbt version per project using `project/build.properties` file (typically maintained by the library authors). This is analogous to `gradlew` for Gradle, Bazelisk for Bazel etc.

Also at the dependency level sbt does use native code regularly. Some of the details are in https://github.com/sbt/sbt/discussions/6202, but whenever we need to deal with new CPU architecture on big operating systems we do sometimes delve into C code.

**Reply by bastien-roucaries** [reply]
> Something to be aware is accidental/historical conflation around what "sbt" refers to.
> 
>     1. sbt can refer to a launcher program named `sbt-launch.jar` (we call this "launcher"). When package managers "install sbt", they aren't actually installing real sbt, they are mostly installing the launcher. This can be downloaded from https://repo1.maven.org/maven2/org/scala-sbt/sbt-launch/1.8.2/sbt-launch-1.8.2.jar, but you can grab it from https://github.com/sbt/sbt/releases/download/v1.8.2/sbt-1.8.2.tgz.

Ok and all the binary are self contained here ? Does the jar run on all arch (like s390x ?)
>     2. sbt can refer to a shell script named `sbt` (we call this "runner"), which calls the launcher, and also included into https://github.com/sbt/sbt/releases/download/v1.8.2/sbt-1.8.2.tgz.
ok take it
> 
>     3. sbt can refer to the program (I call this "mother ship" or "main artifact") that does the building. This is published to Maven Central, and needs to be resolved using some resolver, like the launcher.

Ok understand

> 
> This mechanism allows users to install the runner + launcher machine-wide, but change the sbt version per project using `project/build.properties` file (typically maintained by the library authors). This is analogous to `gradlew` for Gradle, Bazelisk for Bazel etc.
> 
> Also at the dependency level sbt does use native code regularly. Some of the details are in #6202, but whenever we need to deal with new CPU architecture on big operating systems we do sometimes delve into C code.

**Reply by eed3si9n** [reply]
> Ok and all the binary are self contained here ?

No just the runner and launcher. Including a single version of sbt had various issues, including making the zip too big when it may not be used.

> Does the jar run on all arch (like s390x ?

I'm not sure, but I doubt anything runs on all arch.
